xds/internal/balancer/outlierdetection/subconn_wrapper.go

```diff
@@ -31,23 +33,51 @@ import (
 // whether or not this SubConn is ejected.
 type subConnWrapper struct {
 	balancer.SubConn
+	// The following fields are set during object creation and read-only after
+	// that.
+
 	listener func(balancer.SubConnState)
+	// healthListenerEnabled indicates whether the leaf LB policy is using a
+	// generic health listener. When enabled, ejection updates are sent via the
+	// health listener instead of the connectivity listener. Once Dualstack
+	// changes are complete, all SubConns will be created by pickfirst which
+	// uses the health listener.
+	healthListenerEnabled bool
 
 	// addressInfo is a pointer to the subConnWrapper's corresponding address
 	// map entry, if the map entry exists.
 	addressInfo unsafe.Pointer // *addressInfo
+
+	// The following fields are only referenced in the context of a work
+	// serializing buffer and don't need to be protected by a mutex.
+
 	// These two pieces of state will reach eventual consistency due to sync in
 	// run(), and child will always have the correctly updated SubConnState.
-	// latestState is the latest state update from the underlying SubConn. This
-	// is used whenever a SubConn gets unejected.
-	latestState balancer.SubConnState
-	ejected     bool
+
+	// stateForUnjection is the latest state update from the underlying
+	// SubConn that was processed through the serializer. This is used whenever
+	// a SubConn gets unejected. This will be the health state if a health
+	// listener is being used, otherwise it will be the connectivity
+	// state.
+	stateForUnjection balancer.SubConnState
```

Comment:
What do you think about the following approach? (it might not work at the end of the day).

- Here in the `subConnWrapper`, we maintain two states: `latestSubConnHealthState` and `latestSubConnRawState`
  - `latestSubConnHealthState` will be written to via a setter provided here, and will be called from `handleSubConnHealthUpdate`
  - `latestSubConnRawState` will be written to via a setter provided here, and will be called from `handleSubConnUpdate`
- `handleSubConnUpdate` will always call the setter to set the raw connectivity state and will always call `b.child.updateSubConnState`
- `handleSubConnHealthUpdate` will always call the setter to set the health state and will always call `b.child.updateSubConnHealthState`
- `handleEjectedUpdate`,  will call a new method on the `subConnWrapper`, which returns the state to use for ejection and the mechanism to propagate that state (i.e. whether via the health listener callback or the connectivity state callback). `handleEjectedUpdate` could look as follows:
  - get the `subConnWrapper` from the update and set the `ejected` field accordingly
  - call `subConnWrapper.ejectionStateAndMechanism` to get the ejection state and how to forward it. The second return value could just be a boolean which indicates if the state needs to be sent through the health listener.
  - based on the second return value, either `child.updateSubConnHealthState` or `child.updateSubConnState` will be called.
  
With this approach, one field (`stateForUnjection`) in the `subConnWrapper` does not mean two different code paths based on whether health listener is enabled or not. Instead, it is easy to follow when each of those state fields is updated, *and* the `subConnWrapper` knows which field to use when ejection is happening.

It is possible that I'm missing something, but wanted to check with you if this sounds alright and what you think about it?

Reply:
I split the variable to tracking ejection state into two, one for raw connectivity and one for health state. The responsibilities for the wrapper, OD policy and the synchronizing policy were becoming unclear because the OD policy was directly reading scWrapper fields. To make things clearer, I moved the handing of the ejection state into the SubConn wrapper and provided methods on the SubConn wrapper to eject and uneject the subconn.